server/api/signup.js

```diff
@@ -1,51 +1,102 @@
-                                            // otherwise identical hashes different
-    Crypto.pbkdf2(req.body.password, salt, 310000, 32, 'sha256', function(err, hashedPassword) {
+  const salt = Crypto.randomBytes(16); // salt is required for hashing; it makes
+  // otherwise identical hashes different
+  Crypto.pbkdf2(
+    req.body.password,
+    salt,
+    310000,
+    32,
+    "sha256",
+    async function (err, hashedPassword) {
       if (err) {
-        res.status(500).send(err); 
-        return next(err); 
+        res.status(500).send(err);
+        return;
       }
+
       // insert new user into database
-      db.run('INSERT INTO User (username, hashed_password, salt) VALUES (?, ?, ?)', [
-        req.body.username,
-        hashedPassword,
-        salt
-      ], function(err) {
-        if (err) {
-          // TODO: Send more descriptive errors
-          res.status(500).send(err) 
-          return next(err); 
-        }
-        var user = {
-          id: this.lastID,
-          username: req.body.username
-        };
-        // TODO: After correctly signing up, log the user in
-
-        res.status(200).send(`Successfully signed user ${user.username} up!`)
-      });
-    });
-  };
-  
-  module.exports = signup;
\ No newline at end of file
+      try {
+        const newUser = await insertUser(
+          db,
+          req.body.username,
+          hashedPassword,
+          salt,
+          (user) => {
+            res
+              .status(200)
+              .send(`Successfully signed user ${req.body.username} up!`);
+          }
+        );
+        // TODO (integration): After correctly signing up, log the user in
+      } catch (err) {
+        res
+          .status(500)
+          .send(`Error: Failed to insert new user!\nSpecific error: ${err}`);
+        return;
+      }
+    }
+  );
+}
+
+/**
+ * Inserts a new User into the database `db`. Calls a callback on the new user object afterwards.
+ * @param {Database} db - The database to insert the User into.
+ * @param {String} username - The username of the new User.
+ * @param {Buffer} hashedPassword - The hashed password of the new User. This (hopefully) should be generated by `Crypto.pbkdf2()`.
+ * @param {String} salt - The salt used in hashing the password.
+ * @param {Callback} callback - A callback which handles what happens after inserting the user.
+ *                   Should take in one argument: the `user` object generated from `insertUser`.
+ * @returns Nothing.
+ */
+async function insertUser(db, username, hashedPassword, salt, callback) {
+  await db.run(
+    "INSERT INTO User (username, hashed_password, salt) VALUES (?, ?, ?)",
+    [username, hashedPassword, salt],
+    function (err) {
+      // throw an error if there's an issue
+      if (err) throw err;
+
+      // create a user object, handle it elsewhere
+      // TODO: Perhaps this should also generate a user token at this time?
```

Comment:
I did mean session token; and fair. Even at that point, the login process could probably be started here as well so that we don't have any unnecessary code duplication.

Reply:
all it would take is requiring the `login` callback from `login.js` and calling it at the end, but we have to wait for Connor to finish up testing and merge it into main

Reply:
Not gonna do it in this PR.